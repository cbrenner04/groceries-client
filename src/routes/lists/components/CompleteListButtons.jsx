import React from 'react';
import PropTypes from 'prop-types';

const CompletedListButtons = props => (
  <div className="btn-group float-right" role="group">
    <button
      onClick={() => props.onListRefresh(props.list)}
      className="btn btn-link p-0 mr-3"
      disabled={props.userId !== props.list.owner_id}
      style={{ opacity: props.userId !== props.list.owner_id ? 0.3 : 1 }}
      data-test-id="complete-list-refresh"
    >
      <i className="fa fa-refresh fa-2x text-primary" />
    </button>
    <button
      onClick={() => props.onListDeletion(props.list)}
      className="btn btn-link p-0"
      disabled={props.userId !== props.list.owner_id}
      style={{ opacity: props.userId !== props.list.owner_id ? 0.3 : 1 }}
      data-test-id="complete-list-trash"
    >
      <i className="fa fa-trash fa-2x text-danger" />
    </button>
  </div>
);

CompletedListButtons.propTypes = {
  userId: PropTypes.number.isRequired,
  list: PropTypes.shape({
    id: PropTypes.number.isRequired,
    users_list_id: PropTypes.number,
    owner_id: PropTypes.number,
  }).isRequired,
  onListRefresh: PropTypes.func.isRequired,
  onListDeletion: PropTypes.func.isRequired,
};

export default CompletedListButtons;
