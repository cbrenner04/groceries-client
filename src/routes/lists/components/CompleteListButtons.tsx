import React from 'react';
import { ButtonGroup } from 'react-bootstrap';

import { Refresh, Trash, Merge } from '../../../components/ActionButtons';
import type { IList } from '../../../typings';

interface ICompleteListButtonsProps {
  userId: string;
  list: IList;
  onListRefresh: (list: IList) => void;
  onListDeletion: (list: IList) => void;
  multiSelect: boolean;
  selectedLists: IList[];
  handleMerge: () => void;
  pending: boolean;
}

const CompleteListButtons: React.FC<ICompleteListButtonsProps> = (props) => {
  const userIsOwner = props.userId === props.list.owner_id;

  return (
    <ButtonGroup className="float-end">
      <Refresh
        handleClick={() => props.onListRefresh(props.list)}
        disabled={!userIsOwner || props.pending}
        classes={userIsOwner && !props.pending ? 'list-button-enabled' : 'list-button-disabled'}
        testID="complete-list-refresh"
      />
      {props.multiSelect && props.selectedLists.length > 1 && (
        <Merge handleClick={props.handleMerge} testID="complete-list-merge" disabled={props.pending} />
      )}
      <Trash
        handleClick={() => props.onListDeletion(props.list)}
        testID="complete-list-trash"
        disabled={props.pending}
      />
    </ButtonGroup>
  );
};

export default CompleteListButtons;
