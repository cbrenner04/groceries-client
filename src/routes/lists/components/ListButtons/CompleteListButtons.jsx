import React from 'react';
import PropTypes from 'prop-types';
import { ButtonGroup } from 'react-bootstrap';

import { Refresh, Trash } from '../../../../components/ActionButtons';

function CompletedListButtons(props) {
  const userIsOwner = props.userId === props.list.owner_id;
  return (
    <ButtonGroup className="float-right">
      <Refresh
        handleClick={() => props.onListRefresh(props.list)}
        disabled={!userIsOwner}
        style={{ opacity: userIsOwner ? 1 : 0.3 }}
        data-test-id="complete-list-refresh"
      />
      <Trash
        handleClick={() => props.onListDeletion(props.list)}
        disabled={!userIsOwner}
        style={{ opacity: userIsOwner ? 1 : 0.3 }}
        data-test-id="complete-list-trash"
      />
    </ButtonGroup>
  );
}

CompletedListButtons.propTypes = {
  userId: PropTypes.number.isRequired,
  list: PropTypes.shape({
    owner_id: PropTypes.number.isRequired,
  }).isRequired,
  onListRefresh: PropTypes.func.isRequired,
  onListDeletion: PropTypes.func.isRequired,
};

export default CompletedListButtons;
