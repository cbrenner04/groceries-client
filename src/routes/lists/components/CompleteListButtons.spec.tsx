import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import CompleteListButtons from './CompleteListButtons';

function setup(suppliedProps = {}, listOwnerId = 'id1') {
  const user = userEvent.setup();
  const defaultProps = {
    onListRefresh: jest.fn(),
    onListDeletion: jest.fn(),
    userId: 'id1',
    list: {
      id: 'id1',
      owner_id: listOwnerId,
      name: 'foo',
      type: 'GroceryList',
      created_at: 'some date',
      completed: false,
      refreshed: false,
    },
    multiSelect: false,
    selectedLists: [],
    handleMerge: jest.fn(),
    pending: false,
  };
  const props = { ...defaultProps, ...suppliedProps };
  const component = render(<CompleteListButtons {...props} />);

  return { ...component, props, user };
}

describe('CompleteListButtons', () => {
  it('renders refresh disabled when user is not owner', async () => {
    const { container, findByTestId } = setup({ userId: 'id2' }, 'id1');

    expect(container).toMatchSnapshot();
    expect(await findByTestId('complete-list-refresh')).toBeDisabled();
    expect(await findByTestId('complete-list-refresh')).toHaveClass('list-button-disabled');
  });

  it('renders refresh enabled when user is owner', async () => {
    const { container, findByTestId } = setup({ userId: 'id1' }, 'id1');

    expect(container).toMatchSnapshot();
    expect(await findByTestId('complete-list-refresh')).toBeEnabled();
    expect(await findByTestId('complete-list-refresh')).toHaveClass('list-button-enabled');
  });

  it('calls props.onListRefresh when refresh button is clicked', async () => {
    const { findByTestId, props, user } = setup();

    await user.click(await findByTestId('complete-list-refresh'));

    expect(props.onListRefresh).toHaveBeenCalledWith(props.list);
  });

  it('calls props.onListDeletion when delete button is clicked', async () => {
    const { findByTestId, props, user } = setup();

    await user.click(await findByTestId('complete-list-trash'));

    expect(props.onListDeletion).toHaveBeenCalledWith(props.list);
  });

  it('calls handleMerge when Merge is selected', async () => {
    const { container, findByTestId, props, user } = setup({
      multiSelect: true,
      selectedLists: [
        {
          id: 'id1',
          name: 'foo',
          type: 'GroceryList',
          created_at: 'some date',
          completed: false,
          refreshed: false,
          owner_id: 'id1',
        },
        {
          id: 'id12',
          name: 'bar',
          type: 'GroceryList',
          created_at: 'some date',
          completed: false,
          refreshed: false,
          owner_id: 'id1',
        },
      ],
    });

    expect(container).toMatchSnapshot();

    await user.click(await findByTestId('complete-list-merge'));

    expect(props.handleMerge).toHaveBeenCalled();
  });
});
