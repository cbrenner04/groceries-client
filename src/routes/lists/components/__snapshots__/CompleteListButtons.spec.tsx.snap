// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompleteListButtons calls handleMerge when Merge is selected 1`] = `
<div>
  <div
    class="float-end btn-group"
    role="group"
  >
    <button
      class="p-0 me-3 list-button-enabled btn btn-link"
      data-test-id="complete-list-refresh"
      type="button"
    >
      <i
        class="fa fa-redo fa-2x text-primary"
      />
    </button>
    <button
      class="p-0 me-3 btn btn-link"
      data-test-id="complete-list-merge"
      type="button"
    >
      <i
        class="fa fa-compress-alt fa-2x text-warning"
      />
    </button>
    <button
      class="p-0 btn btn-link"
      data-test-id="complete-list-trash"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;

exports[`CompleteListButtons renders refresh disabled when user is not owner 1`] = `
<div>
  <div
    class="float-end btn-group"
    role="group"
  >
    <button
      class="p-0 me-3 list-button-disabled btn btn-link"
      data-test-id="complete-list-refresh"
      disabled=""
      type="button"
    >
      <i
        class="fa fa-redo fa-2x text-primary"
      />
    </button>
    <button
      class="p-0 btn btn-link"
      data-test-id="complete-list-trash"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;

exports[`CompleteListButtons renders refresh enabled when user is owner 1`] = `
<div>
  <div
    class="float-end btn-group"
    role="group"
  >
    <button
      class="p-0 me-3 list-button-enabled btn btn-link"
      data-test-id="complete-list-refresh"
      type="button"
    >
      <i
        class="fa fa-redo fa-2x text-primary"
      />
    </button>
    <button
      class="p-0 btn btn-link"
      data-test-id="complete-list-trash"
      type="button"
    >
      <i
        class="fa fa-trash fa-2x text-danger"
      />
    </button>
  </div>
</div>
`;
